lib/rpush/client/active_model/fcm/notification.rb

```diff
@@ -0,0 +1,118 @@
+          FCM_PRIORITY_HIGH = Rpush::Client::ActiveModel::Apns::Notification::APNS_PRIORITY_IMMEDIATE
+          FCM_PRIORITY_NORMAL = Rpush::Client::ActiveModel::Apns::Notification::APNS_PRIORITY_CONSERVE_POWER
+          FCM_PRIORITIES = [FCM_PRIORITY_HIGH, FCM_PRIORITY_NORMAL]
+
+          ROOT_NOTIFICATION_KEYS = %w[title body image].freeze
+          ANDROID_NOTIFICATION_KEYS = %w[icon tag color click_action body_loc_key body_loc_args title_loc_key
+                                         title_loc_args channel_id ticker sticky event_time local_only
+                                         default_vibrate_timings default_light_settings vibrate_timings
+                                         visibility notification_count light_settings].freeze
+
+          def self.included(base)
+            base.instance_eval do
+              validates :device_token, presence: true
+              validates :priority, inclusion: { in: FCM_PRIORITIES }, allow_nil: true
+
+              validates_with Rpush::Client::ActiveModel::PayloadDataSizeValidator, limit: 4096
+              validates_with Rpush::Client::ActiveModel::RegistrationIdsCountValidator, limit: 1000
+
+              validates_with Rpush::Client::ActiveModel::Fcm::ExpiryCollapseKeyMutualInclusionValidator
+              validates_with Rpush::Client::ActiveModel::Fcm::NotificationKeysInAllowedListValidator
+            end
+          end
+
+          def payload_data_size
+            multi_json_dump(as_json['message']['data']).bytesize
+          end
+
+          # This is a hack. The schema defines `priority` to be an integer, but FCM expects a string.
+          # But for users of rpush to have an API they might expect (setting priority to `high`, not 10)
+          # we do a little conversion here.
+          def priority=(priority)
+            case priority
+              when 'high', FCM_PRIORITY_HIGH
+                super(FCM_PRIORITY_HIGH)
+              when 'normal', FCM_PRIORITY_NORMAL
+                super(FCM_PRIORITY_NORMAL)
+              else
+                errors.add(:priority, 'must be one of either "normal" or "high"')
+            end
+          end
+
+          def dry_run=(value)
+            fail ArgumentError, 'FCM does not support dry run' if value
+          end
+
+          def mutable_content=(value)
+            fail ArgumentError, 'RPush does not currently support mutable_content for FCM' if value
+          end
+
+          def as_json(options = nil) # rubocop:disable Metrics/PerceivedComplexity
+            json = {
+              'data' => data,
+              'android' => android_config,
+              'token' => device_token
+            }
+            json['content_available'] = content_available if content_available
+            json['notification'] = root_notification if notification
+            { 'message' => json }
+          end
+
+          def android_config
+            json = {
+              'notification' => android_notification,
```

Comment:
don't think we need to set explicitly notification here since original library don't have it and we can set it dynamically when someone creating the object, for some use case devs used to configure all stuff under `data` key, on those cases empty notification object is needed which helps custom rendering of the pushes

Reply:
never mind, it has been fixed, actually prod env was running old daemon process which was killed

Reply:
Setting the notification here can cause issues as data messages would not get handled when the application is running in the background.

Reply:
@muzzamilpulsate @KrakenWagen This was done to try and make migration as easy as possible. My understanding of the api changes google provided ([documentation](https://firebase.google.com/docs/cloud-messaging/migrate-v1#example-targeting-multiple-platforms)) was that previously Google's backend manipulated things to make them Android or Apple friendly, while the new api does not. 
So either when upgrading to Rpush FCM users will need to change their backend payload or we put something like this in place. I don't mind either way, I thought easy upgrade was preferable.

That said, if this will cause issues with data messages that obviously would need to be fixed. What exactly is the issue with data messages? Is it that `default_sound` would be set to `true` in this section? I think everything else would be empty for a data message.

Reply:
The problem here could be that 'notification' is a reserved keyword, so even if this is not setting the root 'notification' key, setting 'android'->'notification' is enough to cause data messages to not be handled in the background.

https://firebase.google.com/docs/cloud-messaging/concept-options#data_messages
> Make sure that you do not use any reserved words in your custom key-value pairs. Reserved words include "from", "notification," "message_type", or any word starting with "google" or "gcm."

Reply:
Setting `notification` here makes sense. It's part of AndroidConfig of the message: https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#AndroidConfig

The link you shared is for the `data` key.